### PR TITLE
Enable CycloneDDS talking with ROS2

### PR DESCRIPTION
**Pull request by hexu33**
Signed-off-by: Xu, He <he.xu@intel.com>

Issue Number: 

Objective of pull request:

## Pull request checklist

Your PR fulfills the following requirements:
- [ ] [Issue](https://github.com/lava-nc/lava/issues) created that explains the change and why it's needed
- [ ] Tests are part of the PR (for bug fixes / features)
- [ ] [Docs](https://github.com/lava-nc/docs) reviewed and added / updated if needed (for bug fixes / features)
- [ ] PR conforms to [Coding Conventions](https://lava-nc.org/developer_guide.html#coding-conventions)
- [ ] [PR applys BSD 3-clause or LGPL2.1+ Licenses](https://lava-nc.org/developer_guide.html#add-a-license) to all code files
- [ ] Lint (`flakeheaven lint src/lava tests/`) and (`bandit -r src/lava/.`) pass locally
- [ ] Build tests (`pytest`) passes locally

## Pull request type

 

 

Please check your PR type:

- [ ] Bugfix
- [ ] Feature
- [ ] Code style update (formatting, renaming)
- [ ] Refactoring (no functional changes, no api changes)
- [ ] Build related changes
- [ ] Documentation changes
- [ ] Other (please describe): 

## What is the current behavior?

-

## What is the new behavior?

-

## Does this introduce a breaking change?

- [ ] Yes
- [ ] No

### CODE REVIEW

File: src/lava/magma/runtime/message_infrastructure/examples/ros2/README.md [hunk 1]
```diff
@@ -18,71 +22,152 @@ ros2 pkg create --build-type ament_cmake DDSMetaData
...
+This example also provides the ROS2 package to communicate with DDS port in `ros_talk_with_dds_cpp` and `ros_talk_with_dds_py` folder. Please follow the [guide](https://docs.ros.org/en/foxy/Tutorials/Beginner-Client-Libraries/Writing-A-Simple-Cpp-Publisher-And-Subscriber.html) to build the cpp code in the folder as a ROS2 package. And for the python code, please follow this [guide](https://docs.ros.org/en/foxy/Tutorials/Beginner-Client-Libraries/Writing-A-Simple-Py-Publisher-And-Subscriber.html).
 ## Build the DDS Example code
 Please follow the `README.md` in path `src/lava/magma/runtime/message_infrastructure/` to build the message infrastrucure library and initial the environment.
 
-Note : Please run the `cmake` command with the options,
+Note : Please run the `cmake` command with the options to choose to build the project with FASTDDS or CycloneDDS.
+```
+cmake .. -DDDS_CHANNEL=ON -DFASTDDS_ENABLE=ON -DCycloneDDS=OFF
```
Comment (killight98) [inline src/lava/magma/runtime/message_infrastructure/examples/ros2/README.md:31]: remove `-DCycloneDDS=OFF`

File: src/lava/magma/runtime/message_infrastructure/examples/ros2/README.md [hunk 2]
```diff
@@ -18,71 +22,152 @@ ros2 pkg create --build-type ament_cmake DDSMetaData
...
+```
+or 
 ```
-cmake .. -DDDS_CHANNEL=ON -DFASTDDS_ENABLE=ON
+cmake .. -DDDS_CHANNEL=ON -DFASTDDS_ENABLE=OFF -DCycloneDDS=ON
```
Comment (killight98) [inline src/lava/magma/runtime/message_infrastructure/examples/ros2/README.md:35]: remove `-DFASTDDS_ENABLE=OFF`

File: src/lava/magma/runtime/message_infrastructure/examples/ros2/README.md [hunk 3]
```diff
@@ -18,71 +22,152 @@ ros2 pkg create --build-type ament_cmake DDSMetaData
...
     $ source /opt/ros/foxy/setup.bash
     $ . install/local_setup.bash
     ```
-3. Enable Qos configuration in `ros_talk_with_dds/profile.xml`. [Configuring Fast DDS in ROS 2](https://fast-dds.docs.eprosima.com/en/latest/fastdds/ros2/ros2_configure.html) could be a reference for understanding the QOS configuration for ROS2.
+3. Enable Qos configuration in `profile.xml`. [Configuring Fast DDS in ROS 2](https://fast-dds.docs.eprosima.com/en/latest/fastdds/ros2/ros2_configure.html) could be a reference for understanding the QOS configuration for ROS2.
 
     Run the commands,
     ```
-    $ export FASTRTPS_DEFAULT_PROFILES_FILE=<lava_repo>/src/lava/magma/runtime/message_infrastructure/examples/ros2/ros_talk_with_dds/profile.xml
+    $ export FASTRTPS_DEFAULT_PROFILES_FILE=<lava_repo>/src/lava/magma/runtime/message_infrastructure/examples/ros2/profile.xml
```
Comment (killight98) [inline src/lava/magma/runtime/message_infrastructure/examples/ros2/README.md:53]: assuming the root dir is message_infastructure, change to 
```
export FASTRTPS_DEFAULT_PROFILES_FILE=<message_infastructure>/examples/ros2/profile.xml
```

File: src/lava/magma/runtime/message_infrastructure/examples/ros2/README.md [hunk 4]
```diff
@@ -18,71 +22,152 @@ ros2 pkg create --build-type ament_cmake DDSMetaData
...
+    python test_fastdds_to_ros.py
+    ```
+2. `test_fastdds_from_ros.py` is corresponding to `publisher` ROS2 node. Please run the command:
+    ```
+    python test_fastdds_from_ros.py
+    ```
+#### CPP Example
+To enbale CPP example, when users are processing the steps in 'Build the DDS Example code' part, the option for build the CPP test example need to be set on, just as,
+```
+cmake .. -DDDS_CHANNEL=ON -DFASTDDS_ENABLE=ON -DCycloneDDS=OFF -DCMAKE_BUILD_TYPE=Debug
```
Comment (killight98) [inline src/lava/magma/runtime/message_infrastructure/examples/ros2/README.md:90]: remove `-DCycloneDDS=OFF`

File: src/lava/magma/runtime/message_infrastructure/examples/ros2/README.md [hunk 5]
```diff
@@ -18,71 +22,152 @@ ros2 pkg create --build-type ament_cmake DDSMetaData
...
+```
+Then the CPP example test binaries will be generated in the folder,
+`<lava_repo>/src/lava/magma/runtime/message_infrastructure/build/test`
```
Comment (killight98) [inline src/lava/magma/runtime/message_infrastructure/examples/ros2/README.md:93]: `<message_infastructure>/build/test`

File: src/lava/magma/runtime/message_infrastructure/examples/ros2/README.md [hunk 6]
```diff
@@ -18,71 +22,152 @@ ros2 pkg create --build-type ament_cmake DDSMetaData
...
+and the names for the tests are `test_fastdds_to_ros` and `test_fastdds_from_ros`.
+1. Navigate to the msg_lib build folder,
+    ```
+    cd <lava_repo>/src/lava/magma/runtime/message_infrastructure/build
```
Comment (killight98) [inline src/lava/magma/runtime/message_infrastructure/examples/ros2/README.md:97]:
```
<message_infastructure>/build/
```

File: src/lava/magma/runtime/message_infrastructure/examples/ros2/README.md [hunk 7]
```diff
@@ -18,71 +22,152 @@ ros2 pkg create --build-type ament_cmake DDSMetaData
...
+2. `test_cyclonedds_from_ros.py` is corresponding to `publisher` ROS2 node. Please run the command:
     ```
-    python test_ros_to_dds.py
+    python test_cyclonedds_from_ros.py
     ```
 #### CPP Example
 To enbale CPP example, when users are processing the steps in 'Build the DDS Example code' part, the option for build the CPP test example need to be set on, just as,
 ```
-cmake .. -DDDS_CHANNEL=ON -DFASTDDS_ENABLE=ON -DCMAKE_BUILD_TYPE=Debug
+cmake .. -DDDS_CHANNEL=ON -DFASTDDS_ENABLE=OFF -DCycloneDDS=ON -DCMAKE_BUILD_TYPE=Debug
```
Comment (killight98) [inline src/lava/magma/runtime/message_infrastructure/examples/ros2/README.md:157]: remove `-DFASTDDS_ENABLE=OFF`

**Review by killight98 — CHANGES_REQUESTED** [review]

File: src/lava/magma/runtime/message_infrastructure/examples/ros2/ros_talk_with_dds_py/ros_talk_with_dds_py/publisher.py [hunk 8]
```diff
@@ -0,0 +1,54 @@
...
+    def __init__(self):
+        super().__init__('minimal_publisher')
+        ros_qos = qos_profile_system_default
+        self.publisher_ = self.create_publisher(DDSMetaData,
+                                                'dds_topic',
+                                                ros_qos)
+        timer_period = 0.5  # seconds
+        self.timer = self.create_timer(timer_period, self.timer_callback)
+        self.i = 0
+        self.msg = DDSMetaData()
```
Comment (killight98) [inline src/lava/magma/runtime/message_infrastructure/examples/ros2/ros_talk_with_dds_py/ros_talk_with_dds_py/publisher.py:23]: could add a function that converts numpy array to DDSMetaData into `utils.py`?

File: src/lava/magma/runtime/message_infrastructure/examples/ros2/ros_talk_with_dds_py/ros_talk_with_dds_py/subscriber.py [hunk 9]
```diff
@@ -0,0 +1,43 @@
...
+        super().__init__('minimal_subscriber')
+        ros_qos = qos_profile_system_default
+        self.subscription = self.create_subscription(
+            DDSMetaData,
+            'dds_topic',
+            self.listener_callback,
+            qos_profile_system_default)
+        self.subscription  # prevent unused variable warning
+
+    def listener_callback(self, msg):
```
Comment (killight98) [inline src/lava/magma/runtime/message_infrastructure/examples/ros2/ros_talk_with_dds_py/ros_talk_with_dds_py/subscriber.py:24]: could add a function that converts DDSMetaData to numpy array into utils.py?

**Review by killight98 — CHANGES_REQUESTED** [review]

File: src/lava/magma/runtime/message_infrastructure/examples/ros2/ros_talk_with_dds_py/ros_talk_with_dds_py/publisher.py (hunk 8 above)
Comment (hexu33) [reply src/lava/magma/runtime/message_infrastructure/examples/ros2/ros_talk_with_dds_py/ros_talk_with_dds_py/publisher.py]: Fix in commit [eca5172](https://github.com/lava-nc/lava/pull/498/commits/eca5172e5d3126b0600e0ec612a80fab0558cf70)

File: src/lava/magma/runtime/message_infrastructure/examples/ros2/ros_talk_with_dds_py/ros_talk_with_dds_py/subscriber.py (hunk 9 above)
Comment (hexu33) [reply src/lava/magma/runtime/message_infrastructure/examples/ros2/ros_talk_with_dds_py/ros_talk_with_dds_py/subscriber.py:25]: Fix in commit [eca5172](https://github.com/lava-nc/lava/pull/498/commits/eca5172e5d3126b0600e0ec612a80fab0558cf70)

File: src/lava/magma/runtime/message_infrastructure/examples/ros2/ros_talk_with_dds_py/ros_talk_with_dds_py/utils/np_mdata_trans.py [hunk 10]
```diff
@@ -0,0 +1,45 @@
+import numpy as np
+from enum import Enum
+from ddsmetadata.msg import DDSMetaData
+
+
+def dtype_num_to_char(num):
+    trans = ['?', 'b', 'B', 'h', 'H', 'l', 'L',
```
Comment (killight98) [inline src/lava/magma/runtime/message_infrastructure/examples/ros2/ros_talk_with_dds_py/ros_talk_with_dds_py/utils/np_mdata_trans.py:7]: replace `trans` by public value `DTYPE_LIST`
```py

DTYPE_LIST = [np.dtype(i) for i in [np.bool ...]]
```

**Review by killight98 — CHANGES_REQUESTED** [review]

**Review by killight98 — APPROVED** [review]

**Pull request merged**